### ISSUE TITLE
Abnormally high CPU usage in Tumblr

### ISSUE THREAD

**Issue by oddPT** [cpriest/SnapLinksPlus#231]
I use 10 Firefox addons, one of those is Snap Link Plus. For the last days I have been experiencing slowdows and huge CPU usage in Tumblr pages, I usually have 20-30 Tumblr pages open and around 50-100 images open in Firefox, spread out in 2 Firefox windows.
Today I decided to test all addons to see if any was misbehaving and causing the slowdowns, or if it was Firefoxs fault. Found the guilty one and it was Snap Link Plus. Even with all other addons disabled Snap Link Plus takes control of one of the Firefox threads and CPU usage stays around 25% (quad-core CPU), it stays like that until I manually terminate Firefox task.
It also hapens when just opening ~60 raw Tumblr images (NSFW) I have bookmarked.

I went back to the old versions available in Firefox Addons page and the last version that works without this issue is 3.1.2.3.

**Comment by oddPT** [comment]
- Only noticed it in Tumblr pages.
- Only when Snap Link Plus is activated. If SLP is the only addon activated, it happens; With all addons activated it happens; If I leave SLP disabled and all other addons activated, it does not happen.
- Havent used it before.
- Just the images, it happens. I havent noticed it happening from other websites.

Have no problem with the 3.1.2.3 version. So it should be something added\removed\altered on 3.1.2.4.

**Comment by cpriest** [comment]
When I say activated, I mean you've used it on a page, not that it's 
just enabled -- I assumed that much. :)

So it happens when just the images are loaded on tumblr?  Is it actually 
the image file or an image page?  (The URL would end with a graphic file 
format e.g. .png/.gif/.jpg...)

Any chance you could find one page that is not NSFW that causes it?

Does it happen with just one or two tublr pages loaded?

**Comment by oddPT** [comment]
- Has not been used before. I restarted Firefoxes several times to pinpoint whats was causing the slowdown.
- It can be a mix, Tumblr pages plus image tabs. Images, one per tab, are the Tumblr raw images.
e.g.
This:
http://data.tumblr.com/55d0bfc4836934c221935bf5313809af/tumblr_mx9jyruNd61sbjygjo1_raw.jpg
Not this:
https://78.media.tumblr.com/55d0bfc4836934c221935bf5313809af/tumblr_mx9jyruNd61sbjygjo1_500.jpg
- Only one page wouldnt trigger it. 
- No.

I went and opened a series of Tumblr images (SFW), a bunch of aviation and military related images. Exported the links to a HTML file so if you want you can download the file, open with Firefox and use Snap Lik Plus to open all the images. Link expires in a week.
https://meocloud.pt/link/80a3f7d8-acc1-4f98-9af4-0c6b3325a5c6/new%201.html/
I tested with the 3.1.2.3 and had no problem, with 3.1.2.4 Firefox stops responding and a Firefox thread stays at ~25% usage.
Hope you can replicate the issue.

**Comment by cpriest** [comment]
Cool, got the download, will investigate and see what's going on.

**Comment by oddPT** [comment]
Today tried version 3.1.3 , same problem as 3.1.2.4.
Last good version for me is 3.1.2.3.

**Comment by cpriest** [comment]
Unfortunately I haven't had time to deal with this, other priorities.  
The 3.1.2.4 release was from a user-submitted feature patch.

**Comment by kulfoon** [comment]
The issue occurs to me as well, in other websites, as I come from Poland I noticed that the issue occurs already on 2 big polish websites + as I use Notepad++ I noticed that it also occurs on 1 english website. No need to open 20-30 tabs in my case, just press CTRL + ALT + DEL to open the windows task manager and then just click the link and visit a single webpage to notice that the whole cpu's single core is utilised at 100% by the browser (it's being showed as 25% as It's a 4 core cpu) + there are also memory usage fluctuations:

https://www.dobreprogramy.pl/

<details not-open>
<summary>Animation 1</summary>

![Anim 001_dobreprogramy](https://user-images.githubusercontent.com/43844449/54474882-f3d12480-47ea-11e9-9a33-467568ec2297.gif)

</details>

https://www.nerwica.com/

<details not-open>
<summary>Animation 2</summary>

![Anim 002_nerwica](https://user-images.githubusercontent.com/43844449/54474883-f469bb00-47ea-11e9-8713-9e6ef7fe1dfc.gif)

</details>

https://notepad-plus-plus.org/community/

<details not-open>
<summary>Animation 3</summary>

![Anim 003 notepad](https://user-images.githubusercontent.com/43844449/54474884-f469bb00-47ea-11e9-976c-8813208480d8.gif)


</details>

I don't know if this will help but I used adblocker (either uBlock Origin or AdBlock Plus) to determine which the website's libraries are in the interference with Snap Links Plus, so after adding the rules:
`||std.wpcdn.pl/wpjslib/wpjslib-sgap.js$script,domain=www.nerwica.com|www.dobreprogramy.pl`
`||notepad-plus-plus.org/community/nodebb.min.js$script,~third-party`
the issue is gone and temporarily fixed at the cost of damage to the functionality of the sites (as these are major the website's scripts, some functionalities of the websites most probably gets broken).

As for Tumblr / Reddit I'm not sure as there's no exact links provided in the thread, if anyone gives me a link(s) I can investigate then.

- Firefox 65.0.2 64bit & Snap Links Plus 3.1.5
- downgrading back to Snap Links Plus 3.1.1.3 fixes the issue (perhaps 3.1.2.3 too but can't find it anymore)
- occurs even on a different browser profile with all addons disabled and only Snap Links Plus enabled, after disabling Snap Links Plus, the issue is gone
- couldn't reproduce the issue on Opera 58 64bit with SNap Links 3.1.4

**Comment by kulmegil** [comment]
My best guess: it's connected with "src/lib/js-csp.min.js" dependency.
If all other content scripts are removed except that one - it will still cause High CPU Impact.

**Comment by cpriest** [comment]
That's a good possibility however with that piece missing the plugin would not work.

For the upcoming 3.1.7 release (out any day now), I've taken on this startup performance problem which has included writing my own js-csp (much simpler) and my own storage handler instead of the library I had been using.  I believe things will be much faster.  

I'll close this at that time, but if your use cases do not show speed improvements, please post a message here about it.

**Comment by kulmegil** [comment]
I'm not sure about what the "startup pefrormace" problem is. Just to be clear, the main issue is that extension causes loop in JS on affected sites - traced to the mentioned lib. Not sure about details since site code is minified. Opening a few affected tabs will cause 100% CPU usage. 

BUT I did build 3.1.7 and it appears removing the old js-csp dependency was the right choice. I opened all 39 tabs with affected domains from #284 at once and CPU usage was on expected level.

So... good rewrite !
